Factorize validators

hectorgomezv opened the pull request:
### Problem: 
All the validators have the same code, they apply the ad hoc validation function and collect/throw the errors:

```javascript
if (!this.isValidWhatever(data)) {
  const errors = this.isValidWhatever.errors as DefinedError[];
  throw this.validationErrorFactory.from(errors);
}
```

We also have to implement the same tests for every validator. This generates a lot of duplicated code with zero benefits. 

### Proposed solution:
This PR aims to simplify a bit the validation logic by introducing a `SimpleValidator` (name open to suggestions, other options could be `GenericValidator` or `ItemValidator`) which encapsulates the common code and tests. This makes creating a new validator an even simpler process: the developer just have create an AJV schema, build the related `ValidateFunction` and call this `SimpleValidator`.

fmrsabino commented:
> We also have to implement the same tests for every validator. This generates a lot of duplicated code with zero benefits.

Curious about the reason you mention this. We should treat validators as a black-box and every entity should have different ways to be validated. Therefore the tests should not be the same as we should test with different inputs (json payloads) and check if the validator is working or not.

moisses89 commented:
> > We also have to implement the same tests for every validator. This generates a lot of duplicated code with zero benefits.
> 
> Curious about the reason you mention this. We should treat validators as a black-box and every entity should have different ways to be validated. Therefore the tests should not be the same as we should test with different inputs (json payloads) and check if the validator is working or not.

I think that the idea here is group the common validators, because it's true that now we have repeated code, the unique difference is the schema that we use.
I would prefer `GenericValidator`. 
Regarding the tests, I agree with @fmrsabino I would prefer test every schema.

hectorgomezv commented:
> > We also have to implement the same tests for every validator. This generates a lot of duplicated code with zero benefits.
> 
> Curious about the reason you mention this. We should treat validators as a black-box and every entity should have different ways to be validated. Therefore the tests should not be the same as we should test with different inputs (json payloads) and check if the validator is working or not.

Hmm, I kind of agree we have to see the validators as black boxes, but I think we don't have to test them with different inputs. I'd say we could test they use the expected schema (something that could be added to this PR) but we don't need to test that the validator behaves as we expect for every input payload combination. 

For me that would be entering into a rabbit hole where we have to test a multiple input combinations and, in the end, we would be testing AJV itself.

hectorgomezv commented:
>  Regarding the tests, I agree with @fmrsabino I would prefer test every schema.

I did this in the past and I wasn't happy at all with it. At the end you are testing that AJV (Joy or whatever validation library you're using) works. That should be covered by AJV tests, not ours. Imagine implementing a persistence layer where you define persistence models and you test the DB throws an error when you send the incorrect data type. Which is the value of those tests? For me this is the same.

hectorgomezv commented:
I added a test to every validator to check the correct schema is used in [3234b5ac](https://github.com/5afe/safe-client-gateway-nest/pull/122/commits/3234b5aca77dcfac095e19c2b5c19a3922542a54)
I renamed `SimpleValidator` to `GenericValidator` ad @moisses89 suggested 👍 ([153d146](https://github.com/5afe/safe-client-gateway-nest/pull/122/commits/153d146313e661ddce27a1f01dd0d7b852916f54))

fmrsabino reviewed (changes requested):
I'm a little bit concerned that we might be going with an over-engineered solution with the validators. By removing the pattern of throwing when we have an error we introduced another pattern where we have a validator per domain object which just calls the `generic.validator.ts` with the correct types.

My suggestion is to evaluate the need of having a validator per domain class – what would be the disadvantage of using the `GenericValidator` in the Repositories given that the GenericValidator can already have the type information associated with it? (with my suggestion commented below).

fmrsabino commented:
> Hmm, I kind of agree we have to see the validators as black boxes, but I think we don't have to test them with different inputs. I'd say we could test they use the expected schema (something that could be added to this PR) but we don't need to test that the validator behaves as we expect for every input payload combination.
> 
> For me that would be entering into a rabbit hole where we have to test a multiple input combinations and, in the end, we would be testing AJV itself.

I disagree that we would be testing AJV with this. Testing AJV would mean that given the schema that we provide it generates correct validation functions for the schema.

We would instead be testing if our validation rules (AJV or not) work as expected:

```
1. Given Payload X
2. Validate X (can be AJV or any other solution)
3. Is X valid or not
```

The goal of this testing scenario is not to test AJV but to test that we have the correct validation rules in place. Because we know what is successful and what is an error case. Testing each combination is however quite cumbersome and we'd need some kind of factory to generate multiple distinct payloads.

If we replace AJV with any other solution and while migrating we change one of the rules by mistake we wouldn't be able to catch that rule change.

However, we can cover some of these scenarios on the bigger Unit tests that we have that cover every layer:

```
1. Given that service X returns Y
2. The CGW returns Z
```

I think in the end we should also re-evaluate the value that we are getting from the tests that we introduce – if they are too tied to the implementation then we will always be rewriting and adjusting these tests as the application grow.

hectorgomezv commented on src/domain/collectibles/collectibles.validator.ts:
```diff
@@ -1,17 +1,17 @@
 import { Injectable } from '@nestjs/common';
-import { DefinedError, ValidateFunction } from 'ajv';
+import { ValidateFunction } from 'ajv';
 import { Collectible } from '../../routes/collectibles/entities/collectible.entity';
```
Changed in [2d63aece19a4bba78](https://github.com/5afe/safe-client-gateway-nest/pull/122/commits/2d63aece19a4bba789ac9a5031a01027b8092d0b)

hectorgomezv commented on src/domain/schema/generic.validator.ts:
```diff
@@ -0,0 +1,17 @@
+    private readonly validationErrorFactory: ValidationErrorFactory,
+  ) {}
+
+  execute(fn: ValidateFunction, data: unknown): void {
```
Changed in [2d63aece19a4bba78](https://github.com/5afe/safe-client-gateway-nest/pull/122/commits/2d63aece19a4bba789ac9a5031a01027b8092d0b)

hectorgomezv commented on src/domain/schema/generic.validator.ts:
```diff
@@ -0,0 +1,17 @@
+      const errors = fn.errors as DefinedError[];
+      throw this.validationErrorFactory.from(errors);
+    }
+  }
```
Great suggestion! Changed in [2d63aece19a4bba78](https://github.com/5afe/safe-client-gateway-nest/pull/122/commits/2d63aece19a4bba789ac9a5031a01027b8092d0b) 🙂

hectorgomezv commented:
> > Hmm, I kind of agree we have to see the validators as black boxes, but I think we don't have to test them with different inputs. I'd say we could test they use the expected schema (something that could be added to this PR) but we don't need to test that the validator behaves as we expect for every input payload combination.
> > For me that would be entering into a rabbit hole where we have to test a multiple input combinations and, in the end, we would be testing AJV itself.
> 
> I disagree that we would be testing AJV with this. Testing AJV would mean that given the schema that we provide it generates correct validation functions for the schema.
> 
> We would instead be testing if our validation rules (AJV or not) work as expected

Hmm, I get your point here as well. For me this is a thin line between testing our logic and testing the dependency itself. AJV does more than "given the schema that we provide it generates correct validation functions for the schema", it also actually validates the data using that validation function. So testing the validation function works as the schema tells it to work, seems more a test for AJV. But as I said I get your point, because it could be useful to check the schemas we write are actually accepting/rejecting some random payloads.

I think this could be useful and for me as you said a broader test (controller tests as we have now, or real end-to-end tests in CI in the future) could be a good point to implement this checks.

> ...Testing each combination is however quite cumbersome and we'd need some kind of factory to generate multiple distinct payloads.
Property-based testing tools could help with this I think, but maybe it's too much for a hopefully simple client gateway like this.

> If we replace AJV with any other solution and while migrating we change one of the rules by mistake we wouldn't be able to catch that rule change.
True! Except if the other technology implements JSONSchema as well.
